src/Microsoft.ML.CpuMath/AvxIntrinsics.cs

```diff
@@ -0,0 +1,1654 @@
+                {
+                    Vector256<float> srcVector = Avx.LoadVector256(pSrcCurrent);
+                    Vector256<float> dstVector = Avx.LoadVector256(pDstCurrent);
+                    srcVector = Avx.Multiply(srcVector, scaleVector256);
+                    dstVector = Avx.Add(dstVector, srcVector);
+                    Avx.Store(pResCurrent, dstVector);
+
+                    pSrcCurrent += 8;
+                    pDstCurrent += 8;
+                    pResCurrent += 8;
+                }
+
+                Vector128<float> scaleVector128 = Sse.SetAllVector128(scale);
+
+                while (pResCurrent + 4 <= pResEnd)
+                {
+                    Vector128<float> srcVector = Sse.LoadVector128(pSrcCurrent);
+                    Vector128<float> dstVector = Sse.LoadVector128(pDstCurrent);
+                    srcVector = Sse.Multiply(srcVector, scaleVector128);
+                    dstVector = Sse.Add(dstVector, srcVector);
+                    Sse.Store(pResCurrent, dstVector);
+
+                    pSrcCurrent += 4;
+                    pDstCurrent += 4;
+                    pResCurrent += 4;
+                }
+
+                while (pResCurrent < pResEnd)
+                {
+                    Vector128<float> srcVector = Sse.LoadScalarVector128(pSrcCurrent);
+                    Vector128<float> dstVector = Sse.LoadScalarVector128(pDstCurrent);
+                    srcVector = Sse.MultiplyScalar(srcVector, scaleVector128);
+                    dstVector = Sse.AddScalar(dstVector, srcVector);
+                    Sse.StoreScalar(pResCurrent, dstVector);
+
+                    pSrcCurrent++;
+                    pDstCurrent++;
+                    pResCurrent++;
+                }
+            }
+
+            ZeroUpper();
+        }
+
+        public static unsafe void AddScaleSU(float scale, Span<float> src, Span<int> idx, Span<float> dst)
+        {
+            fixed (float* psrc = src)
+            fixed (int* pidx = idx)
+            fixed (float* pdst = dst)
+            {
+                float* pSrcCurrent = psrc;
+                int* pIdxCurrent = pidx;
+                float* pDstCurrent = pdst;
+                int* pEnd = pidx + idx.Length;
+
+                Vector256<float> scaleVector256 = Avx.SetAllVector256(scale);
+
+                while (pIdxCurrent + 8 <= pEnd)
+                {
+                    Vector256<float> srcVector = Avx.LoadVector256(pSrcCurrent);
+                    Vector256<float> dstVector = Load8(pDstCurrent, pIdxCurrent);
+
+                    srcVector = Avx.Multiply(srcVector, scaleVector256);
+                    dstVector = Avx.Add(dstVector, srcVector);
+                    Store8(in dstVector, pDstCurrent, pIdxCurrent);
+
+                    pIdxCurrent += 8;
+                    pSrcCurrent += 8;
+                }
+
+                Vector128<float> scaleVector128 = Sse.SetAllVector128(scale);
+
+                while (pIdxCurrent + 4 <= pEnd)
```

Comment:
Do we really need a loop here? Since we know we are within 8 elements of the end, we can only do this section at most 1 time.

Reply:
No, we don't need to.  I should probably change all `while` to `if` regarding this case in all methods - thanks!

Meanwhile, I'm looking forward to implementing the 'double-compute' optimization suggested by @tannergooding (the bottom item in https://github.com/briancylui/machinelearning/issues/2), which may drastically change the layout of the intrinsics files and thus will be included in a separate PR.